dask/base.py

```diff
@@ -617,17 +652,27 @@ def compute(
         collections=collections,
         get=get,
     )
+    if newstyle_collections(collections):
+        collections = [c.finalize_compute() for c in collections]
+        graph_factory = collections_to_dsk(collections, optimize_graph, **kwargs)
+        dsk = graph_factory.materialize()
+        keys = graph_factory.__dask_output_keys__()
+        return schedule(dsk, keys, **kwargs)
+    else:
+        graph_factory = collections_to_dsk(collections, optimize_graph, **kwargs)
+        from dask.highlevelgraph import TaskFactoryHLGWrapper
 
-    dsk = collections_to_dsk(collections, optimize_graph, **kwargs)
-    keys, postcomputes = [], []
-    for x in collections:
-        keys.append(x.__dask_keys__())
-        postcomputes.append(x.__dask_postcompute__())
+        assert isinstance(graph_factory, TaskFactoryHLGWrapper)
+        dsk = graph_factory.materialize()
+        keys, postcomputes = [], []
+        for x in collections:
+            keys.append(x.__dask_keys__())
+            postcomputes.append(x.__dask_postcompute__())
 
-    with shorten_traceback():
-        results = schedule(dsk, keys, **kwargs)
+        with shorten_traceback():
+            results = schedule(dsk, keys, **kwargs)
 
-    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
+        return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
 
```

Comment:
the way postcompute actions are defined so far doesn't fit well with the expr/graph factory model since the way it is currently defined, we extract the graph from the collection and mutate it using closures returned by the collection.
I personally consider this indirection very hard to understand and it also doesn't work well since I consider this an abstraction leak since it's a little weird for us having to mutate the graph to compute it. Instead, we tell the collection what it is supposed to be used for (`finalize_compute`) first, then we extract the graph factory, then we generate the graph, then we compute the graph.

I assume it's possible to reduce this boilerplate compat code a little but I haven't put much thought in it, yet.

Reply:
the distributed version is a little more complex (new API much simpler), see https://github.com/dask/distributed/pull/7942/files#r1432859321